### PR TITLE
Address feedback from config binding gen PR to improve enum parsing

### LINKED ISSUE #89879: Configuration source generator should use ArgumentNullException.ThrowIfNull if it's available
If the compilation has `ArgumentNullException.ThrowIfNull` available, it'd be nice if instead of emitting code like:
```C#
public static void BindCore(IConfiguration configuration, ref Tests.MyConfigSection obj, BinderOptions? binderOptions)
{
    if (obj is null)
    {
        throw new ArgumentNullException(nameof(obj));
    }
    ...
```
it emitted code like:
```C#
public static void BindCore(IConfiguration configuration, ref Tests.MyConfigSection obj, BinderOptions? binderOptions)
{
    ArgumentNullException.ThrowIfNull(obj);
    ...
```

### LINKED ISSUE #89936: Address feedback from config binding gen PR to improve enum parsing
Task to address feedback from https://github.com/dotnet/runtime/pull/89823#pullrequestreview-1559404295.

Self-assigning but happy to take a community contribution.

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #89936
Fixes #89879

@layomia I addressed the follow-up fixes.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs [hunk 1]
```diff
@@ -65,6 +65,15 @@ public Parser(SourceProductionContext context, KnownTypeSymbols typeSymbols, Imm
...
 
+                foreach (var typeSymbol in _createdSpecs.Keys)
+                {
+                    if (IsEnum(typeSymbol))
+                    {
+                        _sourceGenSpec.EmitEnumParseMethod = true;
+                        break;
+                    }
+                }
+
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:76]: It would be better to set the property here instead - https://github.com/dotnet/runtime/blob/3fac334c3ab3a61e41a2913ad549e666e91bdeed/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs#L239.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs [hunk 2]
```diff
@@ -128,7 +137,7 @@ private static bool IsValidRootConfigType(ITypeSymbol? type)
                 {
                     ParsableFromStringSpec stringParsableSpec = new(type) { StringParsableTypeKind = specialTypeKind };
 
-                    if (stringParsableSpec.StringParsableTypeKind is not StringParsableTypeKind.AssignFromSectionValue)
+                    if (stringParsableSpec.StringParsableTypeKind != StringParsableTypeKind.AssignFromSectionValue && stringParsableSpec.StringParsableTypeKind != StringParsableTypeKind.Enum)
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:140]: I'd re-write this as 

```suggestion
if (specialTypeKind is StringParsableTypeKind.Enum)
{
    _sourceGenSpec.EmitEnumParseMethod = true;
}
else if (specialTypeKind is not StringParsableTypeKind.AssignFromSectionValue)
{
    _sourceGenSpec.PrimitivesForHelperGen.Add(stringParsableSpec);
}
```

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs (hunk 1 above)
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:76]: Shouldn't need this given the suggestion below.

**Comment by layomia** [comment]
@pedrobsaila, let's go with checking for the right method in the parser - https://github.com/dotnet/runtime/pull/89823#discussion_r1289199719.

**Comment by layomia** [comment]
@pedrobsaila apologies for my delayed response. Could you pls resolve merge conflicts here?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Model/KnownTypeSymbols.cs [hunk 3]
```diff
@@ -108,6 +110,10 @@ public KnownTypeSymbols(CSharpCompilation compilation)
             IReadOnlyList_Unbound = compilation.GetBestTypeByMetadataName(typeof(IReadOnlyList<>))?.ConstructUnboundGenericType();
             IReadOnlySet_Unbound = compilation.GetBestTypeByMetadataName("System.Collections.Generic.IReadOnlySet`1")?.ConstructUnboundGenericType();
             ISet_Unbound = ISet?.ConstructUnboundGenericType();
+
+            // needed to be able to know if a member exist inside the compilation unit
+            Enum = compilation.GetBestTypeByMetadataName(typeof(Enum));
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Model/KnownTypeSymbols.cs:115]: Use the overload taking a `SpecialType` (i.e. `SpecialType.System_Enum`). All the [special types](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.specialtype?view=roslyn-dotnet-4.6.0) are guaranteed to be in the input compilation.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs [hunk 4]
```diff
@@ -63,6 +63,8 @@ public Parser(SourceProductionContext context, KnownTypeSymbols typeSymbols, Imm
                     }
                 }
 
+                _sourceGenSpec.EmitThrowIfNullMethod = !_typeSymbols.ArgumentNullException.GetMembers("ThrowIfNull").IsEmpty;
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:66]:
```suggestion
                _sourceGenSpec.EmitThrowIfNullMethod = _typeSymbols.ArgumentNullException?.GetMembers("ThrowIfNull").IsEmpty is false;
```

The symbol could be null; avoid a null-ref exception.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/ServiceCollection/Configure_T.net462.generated.txt [hunk 5]
```diff
@@ -0,0 +1,243 @@
+// <auto-generated/>
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Baselines/ServiceCollection/Configure_T.net462.generated.txt:1]: Thanks for adding the `net462` variants. Could we refactor to the following structure?

```
> baselines
    > netcoreapp
        > foo.generated.txt
    > net462
        > foo.generated.txt
```

I'd actually suggest making a preliminary PR doing this. We'd need to remove the condition on [[this line]](https://github.com/dotnet/runtime/blob/e80c9aedce4e8d9bdead812cfeedb6c9a2428aec/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj#L46) so that the net462 tests would be included in automated runs. Doing this will reveal a bit of additional, tangential work needed to complete the effort.

**Comment by layomia** [comment]
Oof looks like merging https://github.com/dotnet/runtime/pull/91180 reset your rebase effort. If you take another stab, I'll be sure to check if this PR is merge-ready & queue it up first if so.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs [hunk 6]
```diff
@@ -376,12 +376,9 @@ public class AnotherGraphWithUnsupportedMembers
             string filename,
             string testSourceCode,
             LanguageVersion languageVersion = LanguageVersion.Preview,
-            Action<ImmutableArray<Diagnostic>>? assessDiagnostics = null,
-            ExtensionClassType extType = ExtensionClassType.None)
+            Action<ImmutableArray<Diagnostic>>? assessDiagnostics = null)
         {
-            string path = extType is ExtensionClassType.None
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs:382]: Please use a a preprocessor directive to determine the path

```cs
string path = 
#if NETCOREAPP
    ...
#else 
    ...
    ;
```

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.Baselines.cs [hunk 7]
```diff
@@ -851,7 +852,7 @@ public async Task CollectionsFwk()
         {
             string source = GetCollectionsSource();
 
-            await VerifyAgainstBaselineUsingFile("Collections.net462.generated.txt", source, assessDiagnostics: (d) =>
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.Baselines.cs]: Per https://github.com/dotnet/runtime/pull/89952/commits/f5a2abcfc0dfc5fdb0267c378dd2be745fcda39b#r1321885499, please revert the changes to these invocations.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj [hunk 8]
```diff
@@ -44,10 +93,8 @@
   </ItemGroup>
 
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
-    <Content Include="Baselines\*.generated.txt;
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj:47]: Please revert back to partitioning the baselines based on the binder ext. type. This is helpful when working issues specific to a particular binder.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs [hunk 9]
```diff
@@ -63,7 +63,7 @@ public Parser(SourceProductionContext context, KnownTypeSymbols typeSymbols, Imm
                     }
                 }
 
-                _sourceGenSpec.EmitThrowIfNullMethod = !_typeSymbols.ArgumentNullException.GetMembers("ThrowIfNull").IsEmpty;
+                _sourceGenSpec.EmitThrowIfNullMethod = _typeSymbols.ArgumentNullException is not null && _typeSymbols.ArgumentNullException.GetMembers("ThrowIfNull").IsEmpty is false;
```
Comment (layomia) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:66]: `_sourceGenSpec.EmitThrowIfNullMethod = _typeSymbols.ArgumentNullException?.GetMembers("ThrowIfNull").IsEmpty is false;`

Also -- to be doubly sure, check that the method has the expected parameters.

**Comment by pedrobsaila** [comment]
@layomia could you check it please before the PR become stale again

**Comment by eiriktsarpalis** [comment]
@pedrobsaila apologies for the delayed response, would it be possible to rebase your changes? Thanks!

**Comment by tarekgh** [comment]
@pedrobsaila do you have a chance to resolve the merge conflicts?

**Comment by pedrobsaila** [comment]
I'm on it. The code changed deeply, resolving conflict would take some time

**Comment by pedrobsaila** [comment]
@eiriktsarpalis @tarekgh ready for a new round of reviews

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs [hunk 10]
```diff
@@ -842,6 +848,45 @@ private void RecordDiagnostic(DiagnosticDescriptor descriptor, Location trimmedL
...
             }
+
+            private void CheckIfToEmitParseEnumMethod()
+            {
+                foreach (var typeSymbol in _createdTypeSpecs.Keys)
+                {
+                    if (IsEnum(typeSymbol))
+                    {
+                        _emitEnumParseMethod = true;
+                        _emitGenericParseEnum = _typeSymbols.Enum.GetMembers("Parse").Any(m => m is IMethodSymbol methodSymbol && methodSymbol.IsGenericMethod);
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:859]: >Parse [](http://example.com/codeflow?start=78&length=5)

do we care which overload we are interested in? or we use all of them?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/Helpers.cs [hunk 11]
```diff
@@ -229,18 +229,30 @@ private void EmitBlankLineIfRequired()
...
                     {
-                        {{returnExpr}};
+                        return;
                     }
                     """);
+                }
+                else
+                {
+                    string throwIfNullExpr = useThrowIfNullMethod
+                    ? $"ArgumentNullException.ThrowIfNull({paramName});"
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/Helpers.cs:246]: >ArgumentNullException [](http://example.com/codeflow?start=24&length=21)

do we need full qualified names here? like global::System.ArgumentNullException?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs [hunk 12]
```diff
@@ -842,6 +848,45 @@ private void RecordDiagnostic(DiagnosticDescriptor descriptor, Location trimmedL
...
+
+                    foreach (var throwIfNullMethod in throwIfNullMethods)
+                    {
+                        if (throwIfNullMethod is IMethodSymbol throwIfNullMethodSymbol && throwIfNullMethodSymbol.IsStatic && throwIfNullMethodSymbol.Parameters.Length == 2)
+                        {
+                            var parameters = throwIfNullMethodSymbol.Parameters;
+                            var firstParam = parameters[0];
+                            var secondParam = parameters[1];
+
+                            if (firstParam.Name == "argument" && firstParam.Type.Equals(_typeSymbols.Object, SymbolEqualityComparer.Default)
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:879]: >firstParam.Type.Equals(_typeSymbols.Object [](http://example.com/codeflow?start=65&length=42)

do we really need to define Object at all. Can't we do a simple check like 

firstParam.Type.SpecialType == SpecialType.System_Object?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs [hunk 13]
```diff
@@ -36,7 +36,7 @@ public async Task LangVersionMustBeCharp12OrHigher(LanguageVersion langVersion)
             Assert.Equal(DiagnosticSeverity.Error, diagnostic.Severity);
         }
 
-        [Fact]
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNetCore))]
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs:39]: >IsNetCore [](http://example.com/codeflow?start=77&length=9)

Why changing this to be Net Core only?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj [hunk 14]
```diff
@@ -48,8 +48,15 @@
     <ProjectReference Include="..\..\gen\Microsoft.Extensions.Configuration.Binder.SourceGeneration.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="true" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
-    <Content Include="Baselines\*.generated.txt;Baselines\ConfigurationBinder\*.generated.txt;Baselines\OptionsBuilder\*.generated.txt;Baselines\ServiceCollection\*.generated.txt">
+  <ItemGroup>
+    <Content Include="Baselines\net462\*.generated.txt;
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj:52]: >"Baselines [](http://example.com/codeflow?start=21&length=10)

do we need to copy specific files? can't we include the whole Baseline folder?

**Comment by tarekgh** [comment]
@pedrobsaila thanks for getting this ready. I added a few questions if you can help with.  otherwise LGTM.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/Helpers.cs (hunk 11 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/Helpers.cs:246]: I see we are emitting `using System` so maybe we don't need the full qualified name.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs [hunk 15]
```diff
@@ -842,6 +848,45 @@ private void RecordDiagnostic(DiagnosticDescriptor descriptor, Location trimmedL
...
+                        return;
+                    }
+                }
+            }
+
+            private bool IsThrowIfNullMethodToBeEmitted()
+            {
+                if (_typeSymbols.ArgumentNullException is not null)
+                {
+                    var throwIfNullMethods = _typeSymbols.ArgumentNullException.GetMembers("ThrowIfNull");
```
[the remark this answers is not in the document]
Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:869]: The method signature is as follows :
```
/// <summary>
/// Get all the members of this symbol that have a particular name.
/// </summary>
/// <returns>An ImmutableArray containing all the members of this symbol with the given name. If there are
/// no members with this name, returns an empty ImmutableArray. Never returns Null.</returns>
ImmutableArray<ISymbol> GetMembers(string name);
```

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/Helpers.cs (hunk 11 above)
Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/Helpers.cs:246]: Exactly, a recent change deleted the need for global imports like this

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs (hunk 13 above)
Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs:39]: In main branch, this test already executes only in .NETCore so I'm not changing anything.


https://github.com/dotnet/runtime/blob/216dd61512311698dd321ce1c4880f9dad14366a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj#L51C3-L57C44

For the reason on why it is excluded while other tests are now getting executed in net462 : it uses the record feature.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs [hunk 16]
```diff
@@ -225,7 +225,7 @@ async Task Test(bool expectOutput)
             }
         }
 
-        [Fact]
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNetCore))]
```
[the remark this answers is not in the document]
Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs:228]: same reason here, uses ImmutableArray which is not supported on net462

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj (hunk 14 above)
Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj]: Include just Baselines does not seem to work, the max I could do is this lightweight version :
```
<Content Include="Baselines\*\*\*.generated.txt;
         Baselines\*\*.generated.txt;">
```

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs (hunk 13 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs:39]: should we move the netcore specific baseline files to a folder like NetCoreApp, and we do the similar idea for NETFX too.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj [hunk 17]
```diff
@@ -49,14 +49,8 @@
...
   <ItemGroup>
-    <Content Include="Baselines\net462\*.generated.txt;
-                      Baselines\net462\ConfigurationBinder\*.generated.txt;
-                      Baselines\net462\OptionsBuilder\*.generated.txt;
-                      Baselines\net462\ServiceCollection\*.generated.txt;
-                      Baselines\netcoreapp\*.generated.txt;
-                      Baselines\netcoreapp\ConfigurationBinder\*.generated.txt;
-                      Baselines\netcoreapp\OptionsBuilder\*.generated.txt;
-                      Baselines\netcoreapp\ServiceCollection\*.generated.txt;">
+    <Content Include="Baselines\*\*\*.generated.txt;
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj:52]: this doesn't look right.

Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj:52]: seems right to me. There qre are 2 patterns :
* Baselines\netcoreapp\OptionsBuilder\*.generated.txt => which would translate to Baselines\*\*\*.generated.txt
* Baselines\netcoreapp\Primitives.txt => which would translate to Baselines\*.generated.txt

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/Microsoft.Extensions.Configuration.Binder.SourceGeneration.Tests.csproj:52]: If we can copy the whole Baseline as I pointed out in https://github.com/dotnet/runtime/pull/89952#discussion_r1353284799 that will be better I guess.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs (hunk 13 above)
Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/SourceGenerationTests/GeneratorTests.cs:39]: not sure if I'm understanding your remark well : We already have different baseline files depending on whether they are in `Baselines\netcoreapp` or `Baselines\net462`.

**Review by tarekgh — APPROVED** [review]
Thanks @pedrobsaila, LGTM.
